## Objective

This notebook performs exploratory data analysis (EDA) on the FLY Initiative datasets.

The goals are:

1. Understand the structure and meaning of each dataset.
2. Identify data quality issues.
3. Explore relationships between student participation and program outcomes.
4. Define requirements for the future data cleaning pipeline and stakeholder dashboard.

## Data Sources

The datasets include:

- Program Scores
- Program Attendance
- Student Demographics
- Program NPS and Volunteer Hours
- Volunteer Shifts and Hours Worked
- Registered Volunteers

The analysis focuses on student outcomes, attendance patterns, program engagement, and demographic differences.

## Import Libraries

In [ ]:
# 1. import libraries

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from pathlib import Path

import warnings
warnings.filterwarnings("ignore")

## Define Path

In [ ]:
# 2. define the path to the data folder
DATA_PATH = Path("/Users/hollyhu/Desktop/F.L.Y dashboards/data/raw_data")

for file in DATA_PATH.iterdir():
    print(file.name)

Brown Student Demographics.xlsx
Brown Volunteer Shifts and Hours Worked.xlsx
Brown Program Attendance Jul2025_Jun2026.csv
Registered volunteers.xlsx
Brown Programs with Volunteer Hours and NPS.xlsx
Brown 2025-2026 Program Scores by Student ID.xlsx


## Load Datasets

In [7]:
# load datasets
# 3.1 programe scores
scores = pd.read_excel(
    DATA_PATH / "Brown 2025-2026 Program Scores by Student ID.xlsx",
    header = 0
)

scores.head()

# 3.2 Program Attendance Dataset

attendance = pd.read_csv(
    DATA_PATH / "Brown Program Attendance Jul2025_Jun2026.csv"
)

attendance.head()

# 3.3 Student Demographics Dataset

demographics = pd.read_excel(
    DATA_PATH / "Brown Student Demographics.xlsx",
    header=0
)

demographics.head()

# 3.4 Programs with Volunteer Hours and NPS Dataset

programs = pd.read_excel(
    DATA_PATH / "Brown Programs with Volunteer Hours and NPS.xlsx",
    header=0
)

programs.head()

# 3.5 Volunteer Shifts and Hours Worked Dataset

volunteer_shifts = pd.read_excel(
    DATA_PATH / "Brown Volunteer Shifts and Hours Worked.xlsx",
    header=0
)

volunteer_shifts.head()

# 3.6 Registered Volunteers Dataset

volunteers = pd.read_excel(
    DATA_PATH / "Registered volunteers.xlsx",
    header=0
)

volunteers.head()



,Primary Affiliation,Volunteer Status,Salutation,First Name,Last Name,Title,Ethnicity,Phone,Contact Record Type,Email,Volunteer Hours,First Volunteer Date,Last Volunteer Date,Account Owner,Account Name,Company Name
0,NaN,Active,NaN,Citizens,Bank Volunteer Associates,NaN,Prefer Not to Answer,4014804901,Volunteer,acastillo@flyinitiative.org,3.0,2026-05-04,2026-05-07,Alexandra Castillo,Citizens Bank Volunteer Associates,NaN
1,NaN,Active,NaN,cheryl,Galloway,NaN,Latinx or Hispanic,6177196972,Volunteer,cheryl.galloway@bcsbmail.com,2.0,2026-03-26,2026-04-10,Default Automation User,cheryl Galloway,Bristol County Savings Bank
2,NaN,Active,NaN,Ann-Marie,Lefebvre,NaN,Caucasian,4015959107,Volunteer,ann-marie.lefebvre@bcsbmail.com,2.0,2026-03-20,2026-03-20,Default Automation User,Ann-Marie Lefebvre,Bristol County Savings Bank
3,NaN,Active,NaN,Sira,Salgado,NaN,Latinx or Hispanic,4013406192,Volunteer,sira.salgado@bcsbmail.com,1.0,2026-03-26,2026-03-26,Default Automation User,Sira Salgado,Bristol County Savings Bank
4,NaN,NaN,NaN,Neil,Parrott,NaN,Other/Unknown,4015889021,Volunteer,neil.parrott@gmail.com,0.0,NaT,NaT,Default Automation User,Neil Parrott,Citizens Bank


## Data Structure Overview

In [8]:
# 4.1 Dataset Summary

datasets = {
    "Program Scores": scores,
    "Program Attendance": attendance,
    "Student Demographics": demographics,
    "Programs NPS Volunteer": programs,
    "Volunteer Shifts": volunteer_shifts,
    "Registered Volunteers": volunteers
}


summary = []

for name, df in datasets.items():
    summary.append({
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1]
    })


dataset_summary = pd.DataFrame(summary)

dataset_summary

,Dataset,Rows,Columns
0,Program Scores,5149,25
1,Program Attendance,89927,9
2,Student Demographics,1231,13
3,Programs NPS Volunteer,238,13
4,Volunteer Shifts,159,9
5,Registered Volunteers,19,16


In [ ]:
# 4.2 Dataset Columns

for name, df in datasets.items():
    print("="*80)
    print(name)
    print(df.columns.tolist())

Program Scores
['Student ID', 'Teacher Id', 'Program Owner: Full Name', 'Assignment', 'Score', 'Attendance Percentage', 'Total Pre-assessment Score', 'Total Score(Pre-assessment)', 'Total Post-assessment Score', 'Pre -Assessment Earning Income Score', 'Post-Assessment Earning Income Score', 'Pre -Assessment Investing Score', 'Post-Assessment Investing Score', 'Pre-Assessment Managing Credit Score', 'Post-Assessment Managing Credit Score', 'Pre-Assessment Saving Score', 'Post-Assessment Saving Score', 'Pre-Assessment Spending Score', 'Post-Assessment Spending Score', 'Pre-Assessment Managing Risk Score', 'Post-Assessment Managing Risk Score', 'Program Code', 'Status', 'Total Score(Post-assessment)', 'Program Name']
Program Attendance
['Program Code', 'Program Owner: Full Name', 'Student ID', 'Created Date', 'Status', 'Attendance Status', 'Volunteer Job: Volunteer Job Name', 'Date', 'Program Name']
Student Demographics
['Student ID', 'Last Name', 'Contact Record Type', 'Class Year', 'Age

In [10]:
# 4.3 Dataset Data Types

for name, df in datasets.items():
    print("="*80)
    print(name)
    display(df.dtypes)

Program Scores


Student ID                                object
Teacher Id                               float64
Program Owner: Full Name                  object
Assignment                                object
Score                                    float64
Attendance Percentage                    float64
Total Pre-assessment Score               float64
Total Score(Pre-assessment)              float64
Total Post-assessment Score              float64
Pre -Assessment Earning Income Score     float64
Post-Assessment Earning Income Score     float64
Pre -Assessment Investing Score          float64
Post-Assessment Investing Score          float64
Pre-Assessment Managing Credit Score     float64
Post-Assessment Managing Credit Score    float64
Pre-Assessment Saving Score              float64
Post-Assessment Saving Score             float64
Pre-Assessment Spending Score            float64
Post-Assessment Spending Score           float64
Pre-Assessment Managing Risk Score       float64
Post-Assessment Mana

Program Attendance


Program Code                         object
Program Owner: Full Name             object
Student ID                           object
Created Date                         object
Status                               object
Attendance Status                    object
Volunteer Job: Volunteer Job Name    object
Date                                 object
Program Name                         object
dtype: object

Student Demographics


Student ID                           float64
Last Name                             object
Contact Record Type                   object
Class Year                           float64
Age Group                             object
Ethnicity                             object
Gender                                object
Language                              object
Household Annual Income               object
Household Size                        object
Parents Educational Degree            object
Created Date                  datetime64[ns]
Student Email                         object
dtype: object

Programs NPS Volunteer


Program Name                         object
Program Description                  object
Program Status                       object
Owner                                object
Start Date                   datetime64[ns]
End Date                     datetime64[ns]
Type Of Program                      object
Number of Units                       int64
Number of Students                    int64
Volunteer Jobs                       object
Volunteer Shifts                      int64
Volunteer Completed Hours           float64
NPS Score                           float64
dtype: object

Volunteer Shifts


Campaign: Program Name            object
ShiftID                           object
Start Date & Time         datetime64[ns]
Volunteer Job Name                object
Contact: Full Name                object
Status                            object
Duration (Hours)                 float64
Total Hours Worked               float64
Display on Website                 int64
dtype: object

Registered Volunteers


Primary Affiliation            float64
Volunteer Status                object
Salutation                     float64
First Name                      object
Last Name                       object
Title                          float64
Ethnicity                       object
Phone                            int64
Contact Record Type             object
Email                           object
Volunteer Hours                float64
First Volunteer Date    datetime64[ns]
Last Volunteer Date     datetime64[ns]
Account Owner                   object
Account Name                    object
Company Name                    object
dtype: object

## Data Structure Validation

In [11]:
# Check student coverage across core datasets

student_id_summary = pd.DataFrame({
    "Dataset": [
        "Program Scores",
        "Program Attendance",
        "Student Demographics"
    ],
    "Unique Students": [
        scores["Student ID"].nunique(),
        attendance["Student ID"].nunique(),
        demographics["Student ID"].nunique()
    ],
    "Total Records": [
        scores.shape[0],
        attendance.shape[0],
        demographics.shape[0]
    ]
})

student_id_summary

,Dataset,Unique Students,Total Records
0,Program Scores,1326,5149
1,Program Attendance,1116,89927
2,Student Demographics,1230,1231


In [12]:
# Inspect Student ID format

print("Sample Student IDs:")
display(scores["Student ID"].astype(str).head(20))


print("Manually created IDs starting with SI:")
scores[
    scores["Student ID"]
    .astype(str)
    .str.startswith("SI")
].shape

Sample Student IDs:


0        nan
1     100500
2     SI1730
3     100501
4     SI1732
5     100017
6     100068
7     SI1439
8     100016
9     100040
10    100042
11    100082
12    101096
13    100038
14    100148
15    100147
16    100018
17    101275
18    101276
19    101279
Name: Student ID, dtype: object

Manually created IDs starting with SI:


(31, 25)

In [13]:
# Check assignment/module distribution

scores["Assignment"].value_counts()

Assignment
Behavioral Economics    409
Types of credit         408
Managing credit         395
Careers                 393
Taxes                   393
Paying for college      385
Checking                383
Savings                 376
Budgeting               370
Investing               364
Insurance               322
Name: count, dtype: int64

In [14]:
# Check student program status

scores["Status"].value_counts()

Status
Completed    3413
Confirmed    1481
Canceled      213
Prospect       19
No-Show        11
Name: count, dtype: int64

In [15]:
# Check attendance status distribution

attendance["Attendance Status"].value_counts()

Attendance Status
No Class              59130
P                     19978
A                      9457
T                      1246
Pending Attendance      116
Name: count, dtype: int64

- Program Scores contains 5,149 records from 1,326 unique students and is the primary dataset for analyzing program outcomes.
- Program Attendance contains 89,927 attendance records from 1,116 students and requires aggregation before joining with student performance data.
- Student Demographics provides student-level background information for subgroup and equity analysis.

- Student ID is the main identifier across student-related datasets. It contains both numerical IDs and manually created IDs starting with "SI", so it should be treated as a string variable.

- Program Scores contains key outcome variables, including module scores, attendance percentage, and pre/post assessment scores, which will be used to evaluate program effectiveness.

## Outcome EDA


In [16]:
# Summary statistics for assessment scores

assessment_cols = [
    "Total Score(Pre-assessment)",
    "Total Score(Post-assessment)"
]

scores[assessment_cols].describe()

,Total Score(Pre-assessment),Total Score(Post-assessment)
count,5137.000000,5137.000000
mean,19.591937,22.243119
std,21.939484,32.272257
min,0.000000,0.000000
25%,0.000000,0.000000
50%,13.330000,0.000000
75%,33.330000,46.670000
max,100.000000,100.000000


In [17]:
# Calculate score improvement for exploration

scores["Assessment Improvement"] = (
    scores["Total Score(Post-assessment)"]
    -
    scores["Total Score(Pre-assessment)"]
)

scores["Assessment Improvement"].describe()

count    5137.000000
mean        2.651182
std        30.156867
min      -100.000000
25%       -10.000000
50%         0.000000
75%         6.670000
max       100.000000
Name: Assessment Improvement, dtype: float64

In [18]:
# Explore which modules show higher improvement

module_improvement = (
    scores
    .groupby("Assignment")["Assessment Improvement"]
    .agg(
        ["count", "mean", "median"]
    )
    .sort_values(
        "mean",
        ascending=False
    )
)

module_improvement

,count,mean,median
Assignment,,,
Investing,364,6.013516,0.0
Managing credit,395,5.038000,0.0
Types of credit,408,4.901936,0.0
Insurance,322,4.427205,0.0
Savings,376,3.149043,0.0
Budgeting,370,2.780838,0.0
Checking,383,2.468825,0.0
Taxes,393,2.071450,0.0
Paying for college,385,1.997195,0.0


In [19]:
# Attendance percentage distribution

scores["Attendance Percentage"].describe()

count    4917.000000
mean       74.432672
std        23.187579
min         0.000000
25%        63.640000
50%        81.480000
75%        91.670000
max       100.000000
Name: Attendance Percentage, dtype: float64

In [20]:
# Relationship between attendance and learning improvement

scores[
    [
        "Attendance Percentage",
        "Assessment Improvement"
    ]
].corr()

,Attendance Percentage,Assessment Improvement
Attendance Percentage,1.000000,0.170967
Assessment Improvement,0.170967,1.000000


- Pre- and post-assessment scores are available for 5,137 records, indicating sufficient data for evaluating student learning outcomes.

- The average total assessment score increased from 19.59 (pre-assessment) to 22.24 (post-assessment), with an average improvement of 2.65 points.

- Assessment improvement varies across financial literacy modules. Investing, Managing Credit, and Types of Credit show relatively higher average improvements compared with other modules.

- Attendance percentage has an average of 74.43%, with most students showing relatively high participation levels.

- Attendance percentage has a positive but weak correlation with assessment improvement (correlation coefficient = 0.17), suggesting that attendance may have some relationship with learning outcomes, but other factors may also influence improvement.

In [21]:
# Check demographic variables

demographics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1231 entries, 0 to 1230
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Student ID                  1230 non-null   float64       
 1   Last Name                   1231 non-null   object        
 2   Contact Record Type         1231 non-null   object        
 3   Class Year                  1230 non-null   float64       
 4   Age Group                   1101 non-null   object        
 5   Ethnicity                   1231 non-null   object        
 6   Gender                      1128 non-null   object        
 7   Language                    1116 non-null   object        
 8   Household Annual Income     1126 non-null   object        
 9   Household Size              1128 non-null   object        
 10  Parents Educational Degree  1128 non-null   object        
 11  Created Date                1231 non-null   datetime64[n

In [22]:
# Review categorical variables

categorical_cols = [
    "Age Group",
    "Ethnicity",
    "Gender",
    "Language",
    "Parents Educational Degree",
    "Contact Record Type"
]

for col in categorical_cols:
    print("="*50)
    print(col)
    print(demographics[col].value_counts(dropna=False))

Age Group
Age Group
15-18 years old      685
0-14 years old       382
NaN                  130
19 - 21 years old     18
22 - 25 years old      9
25+                    7
Name: count, dtype: int64
Ethnicity
Ethnicity
Latinx or Hispanic             470
African or African American    294
Prefer Not to Answer           133
Two or More                    128
Caucasian                      103
Other/Unknown                   58
Asian                           27
Native American                 18
Name: count, dtype: int64
Gender
Gender
Female                  565
Male                    525
NaN                     103
Prefer Not to Answer     22
Non-Conforming            8
Not Listed                3
Transgender Male          3
Transgender Female        2
Name: count, dtype: int64
Language
Language
English                 966
NaN                     115
Spanish                 102
Prefer Not to Answer     18
Other                    18
Portuguese                5
Arabic                    4


In [23]:
# Review household-related variables

household_cols = [
    "Household Annual Income",
    "Household Size"
]

demographics[household_cols].describe()

,Household Annual Income,Household Size
count,1126,1128
unique,6,4
top,Prefer Not to Answer,2 - 4
freq,692,522


In [24]:
# Missing value overview for demographics

demographics.isnull().sum()

Student ID                      1
Last Name                       0
Contact Record Type             0
Class Year                      1
Age Group                     130
Ethnicity                       0
Gender                        103
Language                      115
Household Annual Income       105
Household Size                103
Parents Educational Degree    103
Created Date                    0
Student Email                 164
dtype: int64

- Student Demographics contains 1,231 records representing 1,230 unique students.

- The dataset provides demographic information including age group, ethnicity, gender, language, household information, and parental education background.

- Several demographic variables contain missing values, especially:
    - Age Group (130 missing values)
    - Gender (103 missing values)
    - Language (115 missing values)
    - Household Annual Income (105 missing values)
    - Parents Educational Degree (103 missing values)

- Some variables include "Prefer Not to Answer" responses, which should be treated separately from missing values during future analysis.

- Age Group is mainly concentrated in students aged 15-18 years old and 0-14 years old.

- The dataset provides useful variables for future equity analysis, including differences in program outcomes across demographic groups.

In [25]:
# Check attendance date range

attendance["Date"].min(), attendance["Date"].max()

('1/1/2026', '9/9/2025')

In [26]:
# Number of attendance records per student

attendance.groupby("Student ID").size().describe()

count    1116.000000
mean       80.379032
std        62.507762
min         1.000000
25%        33.000000
50%        76.000000
75%       103.000000
max       772.000000
dtype: float64

In [27]:
# Attendance status distribution

attendance["Attendance Status"].value_counts(dropna=False)

Attendance Status
No Class              59130
P                     19978
A                      9457
T                      1246
Pending Attendance      116
Name: count, dtype: int64

In [28]:
# Check potential duplicate attendance records

attendance.duplicated(
    subset=[
        "Student ID",
        "Date",
        "Volunteer Job: Volunteer Job Name"
    ]
).sum()

np.int64(6226)

- Program Attendance contains 89,927 attendance records from 1,116 unique students.

- Attendance records are highly granular, with an average of approximately 80 attendance records per student. Therefore, attendance data should be aggregated before being joined with student outcome data.

- Attendance Status contains multiple categories:
    - No Class
    - P (Present)
    - A (Absent)
    - T
    - Pending Attendance

- Potential duplicate attendance records were identified (6,226 records) based on Student ID, Date, and Volunteer Job Name. Duplicate handling rules should consider the Created Date field, as newer records may represent updated attendance information.

- The attendance dataset covers dates from September 2025 to January 2026 based on the current data.